consensus/service/src/api/client_api_service.rs

```diff
@@ -1825,4 +1897,112 @@ mod client_api_tests {
+
+        let tx_manager = MockTxManager::new();
+        let is_serving_fn = Arc::new(|| -> bool { true });
+        let authenticator = AnonymousAuthenticator::default();
+
+        const LRU_CAPACITY: usize = 4096;
+        let tracked_sessions = Arc::new(Mutex::new(LruCache::new(LRU_CAPACITY)));
+
+        let mut config = get_config();
+        // Permit only 3 failed transactions
+        config.tx_failure_limit = limit;
+
+        // Cause the mock enclave to consistently fail each request.
+        consensus_enclave
+            .expect_client_tx_propose()
+            .times(limit as usize)
+            .return_const(Err(EnclaveError::MalformedTx(
+                TransactionValidationError::ContainsSpentKeyImage,
+            )));
+        // Expect a close, since this will be exceeding our limit
+        consensus_enclave
+            .expect_client_close()
+            .times(1)
+            .return_const(Ok(()));
+
+        let instance = ClientApiService::new(
+            config,
+            Arc::new(consensus_enclave),
+            scp_client_value_sender,
+            Arc::new(ledger),
+            Arc::new(tx_manager),
+            Arc::new(MockMintTxManager::new()),
+            is_serving_fn,
+            Arc::new(authenticator),
+            logger,
+            // Clone this, maintaining our own Arc reference into the tracked
+            // sessions structure so that we can inspect it later.
+            tracked_sessions.clone(),
+        );
+
+        // gRPC client and server.
+        let (client, _server) = get_client_server(instance);
+        let message = Message::default();
+
+        for _ in 0..limit {
+            let propose_tx_response = client
+                .client_tx_propose(&message)
+                .expect("Client tx propose error");
+            assert_eq!(
+                propose_tx_response.get_result(),
+                ProposeTxResult::ContainsSpentKeyImage
+            );
+        }
+        // No failed transaction proposals over the limit yet, so the session
+        // shouldn't have been dropped
+        {
+            let tracker = tracked_sessions
+                .lock()
+                .expect("Attempt to lock session-tracking mutex failed.");
+            assert_eq!(tracker.len(), 1);
+            let (_session_id, tracking_data) = tracker.iter().next().unwrap();
+            assert_eq!(tracking_data.tx_proposal_failures.len(), limit as usize);
+        }
+
+        let propose_tx_response = client.client_tx_propose(&message);
+        assert!(propose_tx_response.is_err());
+
+        match propose_tx_response {
+            Err(grpcio::Error::RpcFailure(rpc_status)) => {
+                assert_eq!(rpc_status.code(), RpcStatusCode::RESOURCE_EXHAUSTED);
+            }
+            _ => panic!(
+                "Unexpected response upon continuing to use\
+                        a rate-limited session: {propose_tx_response:?}"
+            ),
+        }
```

Comment:
:wrench: it might lose a little bit of info on failures but perhaps
```suggestion
        assert!(
            matches!(propose_tx_response, Err(grpcio::Error::RpcFailure(failure)) if failure.code() == RpcStatusCode::RESOURCE_EXHAUSTED)
        );

```